### ISSUE TITLE
Add UNIX domain socket transport

### ISSUE THREAD

**Issue by PhrozenByte** [borgbackup/borg#6183]
/kind enhancement

When creating full system backups one must usually run Borg as `root` to avoid permission issues. However, it might be undesirable to store these backups as root, as people often prefer storing backups in known places (like `/var/backups`) owned by dedicated users (like the default `backup` user in Debian-based distributions).

Right now there are only two possible solutions to accomplish this:
- One might `chown` the repo after `borg create` et al. exits
- One might abuse Borg's remote repository feature to connect to `localhost` using SSH as different user (as suggested by @ThomasWaldmann, see references below)

IMO this is a reasonable use case and Borg should support it without taking such rather desperate solutions.

Thus I'd like to suggest adding a `socket://` transport (like `socket:///run/borg/borg.sock`) to Borg. The socket should be created by `borg serve` by accepting an additional `--socket` option with a path (e.g. `borg serve --socket /run/user/1000/borg/borg.sock`, umask 0117). How `borg serve` is being invoked is up to the user. `borg create` et al. can now connect to this socket using the `socket://` transport (e.g. `borg create socket:///run/borg/borg.sock /path/to/backup`).

Even though this is not the goal, it also makes https://github.com/borgbackup/borg/blob/master/docs/deployment/pull-backup.rst#socat easier by eliminating `socat`. It might also enable some more use cases I just can't think of right now.

As I don't know Borg's sources I can't really judge the complexity, but since Borg already supports remote transports with `borg serve`, this could be as easy as replacing `sys.stdin`/`sys.stdout` of `borg serve` with a UNIX domain socket and to let `borg create` et al. communicate with this socket instead of the `ssh` subprocess.

**References**

> For non-ssh repos and running borg as root, but not having the repo owned by root, you can use the ssh://user@localhost/myrepo trick.

\- https://github.com/borgbackup/borg/issues/4082#issuecomment-423679676 

> You can work around that problem by using borg with `user@localhost:/path/to/repo` as repo (in that scenario, borg client can run as root and borg serve would run as user.

\- https://github.com/borgbackup/borg/issues/3587#issuecomment-443694439

**Comment by ThomasWaldmann** [comment]
Sounds good, thanks for the detailled suggestion!

**Comment by ThomasWaldmann** [comment]
The code in `borg.remote` uses FDs (of stdin, stdout, stderr) and `os.select/read/write/set_blocking`.

It seems like `select()` and `setblocking()` can also be used with sockets, so would it be just `send/recv` instead of `write/read? How about error / exception handling?

The remote code is rather complex and not very nice to work with, I'ld like to avoid additional complexity or risks there.

**Comment by RonnyPfannschmidt** [comment]
Sockets can be turned into file objects for feature parity using the makefile api

If those readable /writable objects can be used, it may be possible without pain 

The only caveat would be that the socket server would have a one client at a time limit for lock parity

**Comment by ThomasWaldmann** [comment]
But `socket.makefile` returns a python `file` object, not a OS-level FD (`int`).

**Comment by RonnyPfannschmidt** [comment]
in the case of socket fd integers - on unix read/write are equivalent to send/recv without flags

**Comment by PhrozenByte** [comment]
@ThomasWaldmann You can call `fileno()` on the `file` object to get the fd from a `socket.makefile()` (`Popen.stdin` and `Popen.stdout` are `file` objects, too). As pointed out by @RonnyPfannschmidt, there should be no differences on Unix-like systems. Windows is a whole different story though, but since Windows isn't supported anyway, this should be no major issue. However, I must admit that I personally never used `socket.makefile()`, but only the high-level socket methods...

Thanks to #6270 giving me a nice starting point I just started looking into Borg's remote code and I agree that the current implementation is indeed not so easy to work with. The code very much confused me, especially whether `borg serve` relies on stderr. If `borg serve` relies on stderr, this is going to be a big issue - because with a Unix socket, you just don't have the separation of stdout and stderr, but only a single bidirectional channel. There are solutions, but it isn't as easy as initially thought then...

However, looking at the code, `borg serve` doesn't seem to use stderr. Even though it defines a `stderr_fd`, it only uses it for a single error, but isn't set to stderr anyway, but stdout, and won't be sent over the wire anyway. So, unless I'm missing something, `borg serve` doesn't use stderr.

https://github.com/borgbackup/borg/blob/1e1c92229807f7522a2f7a6aa4bba3bfe46ddbaf/src/borg/remote.py#L198-L201

What confuses me about this is that the "client" part, i.e. the `Popen` call, is differentiating between stdout and stderr:

https://github.com/borgbackup/borg/blob/1e1c92229807f7522a2f7a6aa4bba3bfe46ddbaf/src/borg/remote.py#L589-L592

Popen's stderr is then sent ([here](https://github.com/borgbackup/borg/blob/1e1c92229807f7522a2f7a6aa4bba3bfe46ddbaf/src/borg/remote.py#L892-L906)) to [`handle_remote_line()`](https://github.com/borgbackup/borg/blob/1e1c92229807f7522a2f7a6aa4bba3bfe46ddbaf/src/borg/remote.py#L1054), which is doing quite sophisticated things to send this stuff to logging (and logging only, which is good). So, looking at this (and not understanding much more about Borg's code :exclamation:), I presume that some older versions of `borg serve` did in fact use stderr, but no longer. If this is true, `handle_remote_line()` only exists for backwards compatibility reasons and can be removed in Borg 2.0, because Borg 2.0 isn't supporting old clients anyway. All server errors are sent via stdout and the only thing for which stderr is still needed, is to catch error output of `ssh`, e.g. reporting about failed auth.

If all of this is true, it's basically really just replacing the fds on both ends by the fds of a `socket`. The fds behave the same. Also see #6270 and [my comment there](https://github.com/borgbackup/borg/pull/6270#issuecomment-1426341921).

> The only caveat would be that the socket server would have a one client at a time limit for lock parity.

On a important side note: My intention with this suggestion is not making `borg serve` a "real" server accepting multiple clients simultaneously. For now it's totally fine to still require one `borg serve` process (and unix socket) per client, just as before, and therefore only accept a single client at a time. However, just to make this clear, this can be expanded to support multiple clients simultaneously at some later point. It naturally requires threading then, which AFAIK is a bigger topic for Borg, therefore it's intentionally out of the scope of this suggestion. But it's laying some ground work...

However, out of curiosity @RonnyPfannschmidt, why is `socket.makefile()` limiting it to a single client? As I said earlier, I never used `socket.makefile()`, but unless I understand the API docs wrong, we can (and actually must) call this per client and therefore get separate fds per client?

**Comment by RonnyPfannschmidt** [comment]
Makefile isn't, the need for locking limits based on usage patterns

**Comment by ThomasWaldmann** [comment]
@PhrozenByte btw, windows support is coming, @RayyanAnsari works on it (no remote repos yet though).

**Comment by ThomasWaldmann** [comment]
About `Location` specifier for `--repo`:
```
socket:///abs/path/to/borg.sock::/abs/path/to/repo
```
Is that a good, usual syntax or are there better ideas?

We need to transport socket path and repo path via that URL.

**Comment by ThomasWaldmann** [comment]
Thanks for the feedback.

The reason why I chose double-slash + abs_path (thus: 3 slashes) is that this could grow support for TCP sockets (by adding a server name or IP after the double-slash (without that, being empty, it means "local" and thus no TCP but UDS).

**Comment by ThomasWaldmann** [comment]
A more radical approach could be splitting that url into `--remote=SERVER_URL --repo=local_repo_path`.

**Comment by PhrozenByte** [comment]
Yeah, good point. I agree, adding a `--socket` option isn't perfect and adds some inconsistency, but if combined with your idea to specify a default location for the socket, a user will easily understand how it works.

Thinking about it, adding the socket's path to `--repo` just feels a bit weird to me: with `--repo` we're just asking for the repo's location. Locations that are not local paths, must match `<transport>://<host>/<path to repo>`, because the host is undoubtedly part of the location. But a socket's path is no host - it's another path. Adding a `--socket` option with a default value is very easy to explain, but "magic separators" (like `::`, or requiring an absolute path and a `.sock` suffix) are hard to understand. Wasn't this confusion the reason (among the regex getting unmanageable) why repo path and archive names were split up in Borg 2?

Adding a `--socket` option just feels like the best option, especially when combined with a default value. Deriving the default value from the repo's path sounds reasonable. So, we end up with this:

```
borg create --repo=ssh://user@server:port/path/to/borg.repo
borg create --rsh='ssh -i /path/to/id_rsa' --repo=ssh://user@server:port/path/to/borg.repo

borg create --repo=socket://server:port/path/to/borg.repo  # possible future with TCP socket
borg create --repo=socket:///path/to/borg.repo  # UNIX domain socket in default location
borg create --socket=/other/path/to/borg.sock --repo=socket:///path/to/borg.repo  # UNIX domain socket in custom location
```

In the future we might even not just look in the repo's path, but in other default locations as well. I just had a whole bunch of ideas, e.g. `$XDG_RUNTIME_DIR/borg.sock` if we ever implement a "full Borg server". But that doesn't really matter right now, by keeping the syntax (`<transport>://<host>/<path to repo>`) simple and rather use a separate option, we stay flexible.

> Hmm, how about just mandating that SOCKET:
> 
>     * must have `.sock` suffix
> 
>     * must be an absolute path (start with a slash)
>       ?

In practice, this means that `.sock` becomes a "magic separator", just as "::" would be. IMHO we should avoid this. It's hard to explain and might conflict with otherwise valid paths.

> A more radical approach could be splitting that url into `--remote=SERVER_URL --repo=local_repo_path`.

You mean like `--remote=ssh://server:port --repo=/path/to/borg.repo`? Yeah, pretty radical, but reasonable. `--remote` and `--repo` would then strongly depend on each other. If a user forgets adding `--remote` (e.g. in a wrapper script appending options), Borg might happily do things wrong without even noticing. Plus, we'd still need the `--rsh` option, right? IMO it feels like this would make things harder.

**Comment by ThomasWaldmann** [comment]
@PhrozenByte about stderr: yeah, the borg serve function does not really used it, but the borg repository code emits logging messages on stderr (which usually goes either directly to e.g. the console (for a local repo) or gets sent via the ssh connection to the clientside code in RemoteRepository).

So, guess we have that problem that there is no easy solution with just 1 bidi UDS.

**Comment by PhrozenByte** [comment]
That's very, very unfortunate :unamused:

Since Borg uses Python's built-in logging facility, I'd indeed recommend creating a [`logging.handlers.RemoteHandler`](https://docs.python.org/3/howto/logging.html#handler-basic) that is responsible for sending not just log messages, but also exceptions and errors of `borg serve` (including Borg's repo code) to the client. This handler can and should be used for SSH-based connections, too. To do so we could [hook into `setup_logging()`](https://github.com/borgbackup/borg/blob/9307f804bfc07c7807dbe09f6d83008cc105fdf5/src/borg/remote.py#L326) (replace `is_serve=False` (and `stream=None`) by `handler=StreamHandler`, so that we can pass `handler=RemoteHandler` for the socket transport) and more or less move [the existing code for exceptions](https://github.com/borgbackup/borg/blob/9307f804bfc07c7807dbe09f6d83008cc105fdf5/src/borg/remote.py#L246-L312) to this new `logging.handlers.RemoteHandler`. Additional work is required to handle the log messages on the client side, which definitely requires diving deep into the remote code... I can totally understand that this isn't desirable.

To avoid this we could multiplex stdout and stderr, by e.g. prefixing all messages with a channel indicator. This won't save us any work on the server side (i.e. `borg serve`) because we must still hook into the logging facility (i.e. add a `logging.handlers.RemoteHandler`) to change where resp. how log messages are handled. But it saves a lot work on the client side, because checking for the channel indicator can be done very early in processing. However, please note that this will break any attempt to support older clients. For future clients it won't be much of a problem apart from having a possibly unused channel indicator at the beginning of every message.

Yet another solution is to quite literally open a second channel for stderr (i.e. calling `socket.connect()` resp. `socket.accept()` twice and using the second connection for stderr). It's by far the easiest solution, doesn't require much work on both sides, however, it has the major disadvantage of being incompatible with a possible future TCP-based transport (connecting twice is fine for a unix socket, but not for TCP) and creates future BC issues in general (abolishing the second connection would be a BC-breaking change).

I'd vote for the second approach, it's a good compromise. But indeed, this idea is no longer as easy as I initially thought it would be :disappointed: 

> `logging.handlers.SocketHandler` maybe? Hmm, maybe rather not - that would need the borg client being the "log server" side (maybe not a big issue for UDS, but for TCP socket).

I'm not 100% sure what you mean. Are you referring to some log messages actually belonging to the server, not the client? I agree, this is a delicate topic, but there's no solution other than checking every single log message whether it's meant for the client, or the server. However, due to our "one client, one socket" policy, we don't really have to think about this right now: we just send everything to the client. How we do this (see above), doesn't really matter IMHO. We won't really need this differentiation until we implement a "full Borg server".

**Comment by ThomasWaldmann** [comment]
about socket file **in** the repo directory:

that would limit operations to existing repos, we can't have the socket file in a repo dir that does not yet exist.
so guess we rather forget about that.

**Comment by PhrozenByte** [comment]
Great work Thomas! :clap: :heart: 

I agree. However, the issue what path to use for `--socket` by default remains an issue tough...

The "right place" for the socket is `$XDG_RUNTIME_DIR` for unprivileged users, e.g. `$XDG_RUNTIME_DIR/borg/borg.sock` (let's reserve a directory, who knows for what we might wanna use it in the future...), and `/run/borg/borg.sock` for root.

However, there are two open questions with `$XDG_RUNTIME_DIR`:

1. What to do when `$XDG_RUNTIME_DIR` is not set
2. What to do when `$BORG_BASE_DIR` is set

First, what's the right place for the socket if `$XDG_RUNTIME_DIR` is not set? The [XDG basedir spec](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) doesn't really specify a default location for `$XDG_RUNTIME_DIR`, but Systemd basically made `/run/user/$UID` the default. We might do the same. However, if `/run/user/$UID` doesn't exist yet, we probably won't be able to create it due to insufficient permissions; thus I'd vote against `/run/user/$UID` as "implicit" default if `$XDG_RUNTIME_DIR` is not set and rather use a "safe" fallback. Again, the XDG basedir spec doesn't really tell us what to do, just to use "a replacement directory with similar capabilities". Some recommend `/tmp/run`, but IMHO this is a rather weird alternative... I'd rather recommend falling back to `$BORG_CACHE_DIR/run`.

Second, what shall we do when `$BORG_BASE_DIR` is set? AFAIK this variable was created before Borg followed the XDG basedir specs. Since `$BORG_BASE_DIR` plays pretty nice with the XDG basedir specs for configs and caches, there was no reason not to keep it. But the runtime dir is different. If we assume `$BORG_BASE_DIR` to be the definitive "I want all my stuff there" answer of the user, I'd say that if `$BORG_BASE_DIR` is set, we always use `$BORG_CACHE_DIR/run`, no matter whether `$XDG_RUNTIME_DIR` is set or not, for both unprivileged users and root.

Last, we should allow the user to give us a "definitive answer" where one wants his stuff... So a `$BORG_RUNTIME_DIR` variable should be added, too.

Just for record, I was also thinking about a `$BORG_SOCKET_DIR` variable, but would recommend to be less specific (i.e. better just use `$BORG_RUNTIME_DIR`): If we - some day, maybe never - add a "full Borg server", we will need a location for other runtime data, e.g. the daemon's PID. Adding yet another env variable for a single file is just overkill...

So, let's wrap this up: We always use `$BORG_RUNTIME_DIR/borg.sock`. If `$BORG_RUNTIME_DIR` is set, use it. If it is not set, but `$BORG_BASE_DIR` is set, set `$BORG_RUNTIME_DIR` to `$BORG_CACHE_DIR/run` and use this location. If neither variables are set, we differentiate between root and unprivileged users: For root, set `$BORG_RUNTIME_DIR` to `/run/borg` and use this location. For unprivileged users, set `$BORG_RUNTIME_DIR` to `$XDG_RUNTIME_DIR/borg` if `$XDG_RUNTIME_DIR` is set, otherwise set it to `$BORG_CACHE_DIR/run` and use this location.

**Comment by ThomasWaldmann** [comment]
socket file permissions: due to `constants.UMASK_DEFAULT == 0o0077`, the socket file has `0o700` permissions (user only, no permissions for group).

Would that be a problem? Usually we have these scenarios:
- run borg client and server as same user: no problem
- run borg client as root and server as non-root: no problem either

**Comment by PhrozenByte** [comment]
Yes, group r+w is required for scenarios with different unprivileged users on the same machine; this includes my original scenario in which the `backup` user (or maybe a distinct `borg` user) runs `borg serve`. Using `umask 007` for the socket must be documented though; users must be aware that by using sockets, all members of the group `borg serve` is run with gain access to repos (and one might run `borg serve` with a group different from the primary group). For these scenarios one must always pass `--socket` (note: not for all scenarios though, e.g. for a proxied socket one won't necessarily need `--socket`). However, I don't think that this is an issue, because such setups always require manual setup anyway (that's what I meant with "How `borg serve` is being invoked is up to the user")...

About using `user_runtime_dir()` from the `platformdirs` package: It looks like that this lib works fine for unprivileged users, but yields `/run/user/0` for root. This is a rather unorthodox location. As common practice it should be just `/run` for root.

**Comment by ThomasWaldmann** [comment]
@PhrozenByte could you file a bug at `platformdirs`?

**Comment by PhrozenByte** [comment]
> @PhrozenByte could you file a bug at `platformdirs`?

ref https://github.com/platformdirs/platformdirs/issues/188

**Comment by ThomasWaldmann** [comment]
@PhrozenByte merged stuff into master.

if you have time, guess it would be good to give it some practical testing.
maybe the long running `borg serve --socket` shows some yet unknown issues.
